crates/simulations/src/cozy/agents/supplier.rs

```diff
@@ -0,0 +1,178 @@
+    ) -> Self {
+        Self {
+            _name,
+            address,
+            protocol,
+            set,
+            preferences,
+            rng,
+        }
+    }
+}
+
+impl Agent<CozyUpdate, CozyWorld> for Supplier {
+    fn address(&self) -> Address {
+        self.address
+    }
+
+    fn activation_step(&mut self, state: &mut State<CozyUpdate, CozyWorld>) {
+        let router_approve_tx = self
+            .set
+            .base_token
+            .approve(self.protocol.cozy_router.address(), U256::MAX);
+        let _ = state.execute_evm_tx_and_decode(self.address, router_approve_tx);
+    }
+
+    fn step(
+        &mut self,
+        state: &State<CozyUpdate, CozyWorld>,
+        channel: &AgentChannelSender<CozyUpdate>,
+    ) {
+        if !self
+            .preferences
+            .reaction_time
+            .time_to_react(state.timestamp(), &mut self.rng)
+        {
+            return;
+        }
+
+        // Get current balance.
+        let balance = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.base_token.balance_of(self.address.into()),
+            )
+            .expect("Error getting balance.");
+        let current_set_shares = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.set.balance_of_matured(self.address.into()),
+            )
+            .expect("Error getting shares.");
+        let current_value = state
+            .call_evm_tx_and_decode(
+                self.address,
+                self.set.set.convert_to_assets(current_set_shares),
+            )
+            .expect("Error converting shares to assets.");
+        self.preferences.total_wealth = balance + current_value;
+
+        // Compute optimal allocation.
+        let optimal_allocation = self.compute_optimal_allocation(state);
+        let optimal_allocation_value =
+            f64_to_u256(optimal_allocation) * self.preferences.total_wealth;
+
+        // Supply or withdraw to target optimal allocation.
+        match optimal_allocation_value.cmp(&current_value) {
+            std::cmp::Ordering::Equal => (),
+            std::cmp::Ordering::Greater => {
+                let supply_amount = optimal_allocation_value - current_value;
+                let router_supply_tx = self.protocol.cozy_router.deposit(
+                    self.set.set.address(),
+                    supply_amount,
+                    self.address.into(),
+                    U256::zero(),
+                );
+                log::info!(
+                    "Supplier {} is deposting {} assets.",
+                    self.address,
+                    supply_amount
+                );
+                channel.execute_evm_tx(router_supply_tx);
+            }
+            std::cmp::Ordering::Less => {
+                let withdraw_amount = current_value - optimal_allocation_value;
+                let router_withdraw_tx = self.protocol.cozy_router.withdraw(
```

Comment:
do you need to completeWithdraw after the withdraw delay has elapsed?

Reply:
Ahh, you're right. Fixing.